### PR TITLE
Implement Subscriptions

### LINKED ISSUE #91: Implement Subscriptions for comments
https://github.com/lml/commontator

Check the documentation and add the functionality of subscriptions to comments.

### PR DESCRIPTION
**Pull request by Amansingla97**
Fixes #91

### CODE REVIEW

File: config/environments/development.rb [hunk 1]
```diff
@@ -52,7 +52,7 @@
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
 
   config.active_job.queue_adapter = :sidekiq
-
+  config.action_mailer.default_url_options = { host: "www.circuitverse.org" }
```
Comment (tachyons) [inline config/environments/development.rb]: this should be localhost in development

**Comment by tachyons** [comment]
Looks good, Few nitpicks
1. Add screenshots
2. Check the feasibility of moving logic out of the view using presenter pattern or decorator pattern
3. Check if we can simplify the frontend operations with https://stimulusjs.org/

File: config/environments/development.rb (hunk 1 above)
Comment (Amansingla97) [reply config/environments/development.rb]: yes ill change it

**Comment by Amansingla97** [comment]
Screenshot 2019-03-12 at 3 17 19 PM
Screenshot 2019-03-12 at 3 17 24 PM

**Comment by satu0king** [comment]
Thanks for the PR. Can you summarize what is the current behavior? I noticed that you generated views from the default commentator plugin, can you summarize the changes.

**Comment by Amansingla97** [comment]
I have made small changes to views of subscription which differ from default commentator plugin

->There were syntax error in the orignal repo (I raised a issue in their repo as well)
->We may need to customize the mailer views.So having app/views/commontator/subscriptions_mailer/ is be necessary.
->The UI of overall commontator does not match with our UI.so would have to change it at some point or other.
Rest the working as usual

**Comment by satu0king** [comment]
> I have made small changes to views of subscription which differ from default commentator plugin
> 
> ->There were syntax error's in the orignal repo (I raised a issue in their repo as well)
> ->We may need to customize the mailer views.So having app/views/commontator/subscriptions_mailer/ is necessary.
> ->The UI of overall commontator does not match with our UI.So would have to change it at some point or other in future.
> Rest the working as usual

Thank you, could you also describe how subscriptions happen. Does it happen when the author creates the project, or when someone comments on the project. Or is it a manual subscruption. Also how are the emails sent? Are they sent for every comment? A screenshot for an email will be helpful.

**Comment by Amansingla97** [comment]
->The email happens when someone else comments on the subscribed project.
->The User can opt to manually opt to subscribe.
->Emails are sent via smtp and would be using the same smtp settings.
->When a new comment is generated the email would be sent to all the subscribed users.it will enqueue the tasks in sidekiq which is then processed via smtp.

**Comment by satu0king** [comment]
> ->The email happens when someone else comments on the subscribed project.
> ->The User can opt to manually opt to subscribe.
> ->Emails are sent via smtp and would be using the same smtp settings.
> ->When a new comment is generated the email would be sent to all the subscribed users.it will enqueue the tasks in sidekiq which is then processed via smtp.

Can you configure auto subscription? And manual unsubscription? I really doubt users will ever subscribe even if they would be interested.

**Comment by Amansingla97** [comment]
Screenshot 2019-03-15 at 8 19 30 PM
These were available options , I changed it both automatic and manual.

**Comment by satu0king** [comment]
Is the author of the project automatically subscribed even if he doesn't make a comment?

**Comment by Amansingla97** [comment]
No,I just checked .Commontator does not provide this functionality.

**Comment by satu0king** [comment]
> No,I just checked .Commontator does not provide this functionality.The author would also have to subscribe to comments or automatically when commenting.

I am sure Commontator has some function which we can call to subscribe/unsubscribe? After all, they need to have a way to change it themselves. Simply add a after_create hook?

**Comment by Amansingla97** [comment]
ill check

**Comment by satu0king** [comment]
> ill check

We can resolve this in a different issue.

**Comment by satu0king** [comment]
undefined method `project_url'

This is the error that is coming in production, did you test the mail in development?

**Comment by Amansingla97** [comment]
ya i tested with gem letter_opener
let me see again.I am in class right now.Will definitely look into it.

**Comment by satu0king** [comment]
```
  config.comment_url_proc = ->(comment, app_routes) {
    app_routes.polymorphic_url(comment.thread.commontable, anchor: "comment_#{comment.id}_div")
  }
```
I think the answer is here.
